src/services/ipfs/backend.rs

```diff
@@ -0,0 +1,329 @@
+        let query_string = format!("{}&create=true&truncate=false", self.new_query_string(path));
+
+        self.make_write_request(
+            &query_string,
+            "",
+            path,
+            IpfsRpcV0::Create,
+            Operation::Create,
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn files_read(
+        &self,
+        path: &str,
+        offset: Option<i64>,
+        count: Option<i64>,
+    ) -> io::Result<BytesReader> {
+        let mut query_string = self.new_query_string(path);
+        if offset.is_some() {
+            query_string = format!("{}&offset={}", query_string, offset.unwrap());
+        }
+        if count.is_some() {
+            query_string = format!("{}&count={}", query_string, count.unwrap());
+        }
+
+        let resp = self
+            .make_request(&query_string, IpfsRpcV0::Read, path, Operation::Read)
+            .await?;
+
+        Ok(Box::new(resp.into_body()))
+    }
+
+    pub(crate) async fn files_delete(&self, path: &str) -> io::Result<()> {
+        let query_string = self.new_query_string(path);
+
+        self.make_request(&query_string, IpfsRpcV0::Delete, path, Operation::Delete)
+            .await?;
+        Ok(())
+    }
+
+    pub(crate) async fn files_list(
+        &self,
+        path: &str,
+    ) -> io::Result<isahc::Response<isahc::AsyncBody>> {
+        let query_string = self.new_query_string(path);
+
+        let resp = self
+            .make_request(&query_string, IpfsRpcV0::List, path, Operation::List)
+            .await?;
+        Ok(resp)
+    }
+
+    pub(crate) async fn files_write(
+        &self,
+        path: &str,
+        data: &str,
+    ) -> io::Result<isahc::Response<isahc::AsyncBody>> {
+        let query_string = format!("{}&create=true&truncate=false", self.new_query_string(path));
+
+        let resp = self
+            .make_write_request(
+                &query_string,
+                data,
+                path,
+                IpfsRpcV0::Write,
+                Operation::Write,
+            )
+            .await?;
+
+        Ok(resp)
+    }
+
+    async fn make_request(
+        &self,
+        query_string: &str,
+        action: IpfsRpcV0,
+        path: &str,
+        op: Operation,
+    ) -> io::Result<isahc::Response<isahc::AsyncBody>> {
+        let url = format!("{}{}?{}", self.endpoint, action.to_path(), query_string);
+
+        let req = isahc::Request::post(url.clone());
+        let req = req.body(isahc::AsyncBody::empty()).map_err(|err| {
+            error!("object post: {url} {err:?}");
+            new_request_build_error(op, path, err)
+        })?;
+
+        let resp = self.client.send_async(req).await.map_err(|e| {
+            error!("object {path} {op}: {e:?}");
+            new_request_send_error(op, path, e)
+        })?;
+
+        Ok(resp)
+    }
+
+    async fn make_write_request(
+        &self,
+        query_string: &str,
+        data: &str,
```

Comment:
We must process input data in an async and streaming way to avoid reading all data into memory.

It's OK for this PR. Please raise a new issue for this so that we can address this in the future.